core/iwasm/libraries/libc-builtin/libc_builtin_wrapper.c

```diff
@@ -341,15 +216,29 @@ _vprintf_wa(out_func_t out, void *ctx, const char *fmt, _va_list ap,
 
                     s = start = addr_app_to_native(s_offset);
 
-                    while (*s)
-                        out((int)(*s++), ctx);
+                    str_len = strlen(start);
+                    if (str_len >= UINT32_MAX - 64) {
+                        print_err(out, ctx);
+                        break;
+                    }
+
+                    /* reserve 64 more bytes as there may be width description
+                     * in the fmt */
+                    buf_len = str_len + 64;
 
-                    if (padding == PAD_SPACE_AFTER) {
-                        int remaining = min_width - (int32)(s - start);
-                        while (remaining-- > 0) {
-                            out(' ', ctx);
-                        }
+                    if (!(buf = wasm_runtime_malloc(buf_len))) {
```

Comment:
Can we pre-define a tmp buffer and use it if buf_len is no larger than its size, and allocate memory and buf_len is out of range:
```C
char buf_tmp[64] = { 0 }, *buf = buf_tmp;
..
if (buf_len > sizeof(buf_tmp) {
  if (!(buf = wasm_runtime_malloc(...))) {
...
```
In free:
```C
if (buf != buf_tmp)
  wasm_runtime_free(buf);
```